src/test/java/com/puppycrawl/tools/checkstyle/api/JavadocTokenTypesTest.java

```diff
@@ -74,7 +74,7 @@ public void testTokenValues() {
         assertEquals(msg, 22, JavadocTokenTypes.LITERAL_INCLUDE);
         assertEquals(msg, 39, JavadocTokenTypes.FIELD_NAME);
         assertEquals(msg, 41, JavadocTokenTypes.FIELD_TYPE);
-        assertEquals(msg, 95, JavadocTokenTypes.HTML_TAG_NAME);
+        assertEquals(msg, 96, JavadocTokenTypes.HTML_TAG_NAME);
```

Comment:
@ps-sp , please explain this change. Token values should not change.

Reply:
@romani `OPTGROUP_HTML_TAG_NAME` needs to be before `HTML_TAG_NAME`. Otherwise `<OPTGROUP>` will get recognized as `HTML_TAG_NAME` rather than `OPTGROUP_HTML_TAG_NAME`. ANTLR applies the first rules that matches. Please also have a look at this comment https://github.com/checkstyle/checkstyle/pull/5222#discussion_r146521645

Reply:
@romani Please let me know if there are any discrepancies in the PR or in the comment. Also, let me know if any other change is to be explained further.

Reply:
We can not change Tokens values at all, each change mean bracking compatibility with all custom Checks in this world. All projects with custom Check have to be recompiled again over new version of Checkstyle. If not recompiled , it will result in really weird behavior of Checks... nightnare.
It is too much affect on all users ..... 

Looks like the  only way out is keep values hardcoded in api./JavadocTokenTypes and have build in hardcoded map of ANTLR values to JavadocTokenTypes  values that will be used in [JavadocUtils.getTokenId()](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtils.java#L364) method that is used on creation of [JavadocNodeImpl](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser.java#L328).

Reply:
@romani Compatibility damage results only when any rule/token has to be added in between the existing ones. We must add `OPTGROUP_HTML_TAG_NAME` before `HTML_TAG_NAME` for the reasons I explained above. We can undo this damage by hardcoding the value for `HTML_TAG_NAME`.

Reply:
@Vladlis , please step in, and make investigation on what we can do keep values of token on same values, independently from grammar change

Reply:
@ps-sp @Vladlis If an option doesn't present itself, I recommend posting to ANTLR group who may know more.